crates/ruff_linter/src/rules/flake8_bugbear/rules/suppressed_exception.rs

```diff
@@ -0,0 +1,174 @@
+use ruff_python_ast::{self as ast, ExceptHandler, Expr, Stmt};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::helpers::RaiseStatementVisitor;
+use ruff_python_ast::statement_visitor::StatementVisitor;
+use ruff_python_semantic::Binding;
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for exception with added note not reraised.
+///
+/// ## Why is this bad?
+/// The `.add_note' function does not use the exception.
+/// It will implicitly be equivalent to passing the exception.
+///
+///
+/// ## Example
+/// ```python
+/// try:
+///     print("hello")
+/// except Exception as e:
+///     e.add_note("test")
+/// ```
+///
+/// Use instead:
+/// ```python
+/// try:
+///     print("hello")
+/// except Exception as e:
+///     e.add_note("test")
+///     raise e
+/// ```
+///
+/// ## References
+/// - [Python documentation: `raise`](https://docs.python.org/3/reference/simple_stmts.html#the-raise-statement)
+#[violation]
+pub struct SuppressedException;
+
+impl Violation for SuppressedException {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Caught exception with call to `add_note` not used. Did you forget to `raise` it?")
+    }
+}
+
+#[derive(Debug, Clone)]
+struct AddNote<'a> {
+    receiver: &'a ast::ExprName,
+}
+
+fn match_str_expr(expr: &Expr) -> bool {
+    let Expr::Call(ast::ExprCall { func, .. }) = expr else {
+        return false;
+    };
+
+    let Expr::Name(ast::ExprName { id, .. }) = func.as_ref() else {
+        return false;
+    };
+
+    if *id == "str" {
+        return true;
+    }
+    false
+}
+fn match_add_note(stmt: &Stmt) -> Option<AddNote> {
+    let Stmt::Expr(ast::StmtExpr { value, .. }) = stmt else {
+        return None;
+    };
+
+    let Expr::Call(ast::ExprCall { func, .. }) = value.as_ref() else {
+        return None;
+    };
+
+    let Expr::Attribute(ast::ExprAttribute { value, attr, .. }) = func.as_ref() else {
+        return None;
+    };
+
+    if attr != "add_note" {
+        return None;
+    }
+
+    let Expr::Name(receiver @ ast::ExprName { .. }) = value.as_ref() else {
+        return None;
+    };
+
+    Some(AddNote { receiver })
+}
+
+/// B040
+pub(crate) fn suppressed_exception(checker: &mut Checker, handlers: &[ExceptHandler]) {
+    let mut count_str_references: usize = 0;
+    for handler in handlers {
+        let ExceptHandler::ExceptHandler(ast::ExceptHandlerExceptHandler { body, name, .. }) =
+            handler;
+
+        let raises = {
+            let mut visitor = RaiseStatementVisitor::default();
+            visitor.visit_body(body);
+            visitor.raises
+        };
```

Comment:
I wonder if we can simplify this code by making more use of `Binding`:

* Resolve the `bindings` of `name`
* Iterate over all references of `binding` and 
  * exit if any is statement (use `SemanticModel::statement(binding.expression_id())`) is a a raise statement and the target or cause is the binding node
  * count the `add_note` calls
* Create a diagnostic if there's at least one `add_note` call.

Reply:
According to the upstream implementation, the following code is safe, which requires checking all `raise' statements in scope, not just those with binding 

```
try:
    ...
except Exception as e:
    e.add_note("...")
    raise  # safe
```
On the last point, just checking for the `add_note` call is not enough, because exceptions can be used in the print, function calls etc that are considered to be safe, so I'm counting all references to them. Not sure if there is a better way to find all usages of the exception that are referenced.

Reply:
I'm not sure I follow. Doesn't 

```python
try: ... 
except Exception as e: 
    e.add_note("...") raise # safe
````

introduce a binding for `e`?

Reply:
It does, but my point is that `raise` is not included in the reference list when I filter by exception name.
```
            let bindings: Vec<&Binding> = semantic
                .current_scope()
                .get_all(exception_name.id.as_str())
                .map(|binding_id| semantic.binding(binding_id))
                .filter(|binding| matches!(binding.kind, BindingKind::UnboundException(_)))
                .collect();


            for binding in bindings.iter() {
                for reference_id in binding.references() {
                    let reference = checker.semantic().reference(reference_id);
                    if let Some(node_id) = reference.expression_id() {
                        let stmt = checker.semantic().statement(node_id);
                        
   
                        println!("stmt={:?}", stmt);
                       ...
```

prints just a `add_note` call
```
stmt=Expr(StmtExpr { range: 76..93, value: Call(ExprCall { range: 76..93, func: Attribute(ExprAttribute { range: 76..86, value: Name(ExprName { range: 76..77, id: Name("e"), ctx: Load }), attr: Identifier { id: Name("add_note"), range: 78..86 }, ctx: Load }), arguments: Arguments { range: 86..93, args: [StringLiteral(ExprStringLiteral { range: 87..92, value: StringLiteralValue { inner: Single
```